src/Microsoft.ApplicationInsights/TelemetryClient.cs

```diff
@@ -513,44 +510,94 @@ public void Initialize(ITelemetry telemetry)
             }
 
             telemetry.Context.Initialize(this.Context, instrumentationKey);
-            foreach (ITelemetryInitializer initializer in this.configuration.TelemetryInitializers)
+
+            ISupportAdvancedSampling telemetryWithSampling = telemetry as ISupportAdvancedSampling;
+
+            // Telemetry can be already sampled out if that decision was made before calling Track()
+            bool sampledOut = telemetryWithSampling?.IsProactivelySampledOut ?? false;
+
+            // If telemetry was not sampled out, we need to check its sampling score after we learn its Operation Id
+            bool shoudlCheckSamplingScore = !sampledOut;
+
+            if (!sampledOut)
             {
-                try
+                if (telemetryWithSampling?.SupportsProactiveSampling ?? false)
                 {
-                    initializer.Initialize(telemetry);
+                    for (int index = 0; index < this.configuration.TelemetryInitializers.Count; index++)
+                    {
+                        if (shoudlCheckSamplingScore && !string.IsNullOrEmpty(telemetry.Context.Operation.Id))
+                        {
+                            if (SamplingScoreGenerator.GetSamplingScore(telemetry.Context.Operation.Id) >= this.configuration.GetLastObservedSamplingPercentage(telemetryWithSampling.ItemTypeFlag))
+                            {
+                                telemetryWithSampling.IsProactivelySampledOut = true;
+                                sampledOut = true;
+                                break;
+                            }
+                            else
+                            {
+                                shoudlCheckSamplingScore = false;
+                            }
+                        }
+
+                        try
+                        {
+                            this.configuration.TelemetryInitializers[index].Initialize(telemetry);
+                        }
+                        catch (Exception exception)
+                        {
+                            CoreEventSource.Log.LogError(string.Format(
+                                                            CultureInfo.InvariantCulture,
+                                                            "Exception while initializing {0}, exception message - {1}",
+                                                            this.configuration.TelemetryInitializers[index].GetType().FullName,
+                                                            exception));
+                        }
+                    }                    
                 }
-                catch (Exception exception)
+                else
```

Comment:
i think this `else` can be removed.

I propose introducing a new class `AdvancedSamplingHelper` that does the evaluation above in the IF and could internalize some extra variables.

Reply:
if (!sampledOut) checks for case when telemetry was already sampled out and we do not need to run initializers at all.
if (supportsProactiveSampling) checks whether current telemetry type supports proactive sampling and we may skip initializers after we know its operation ID (if branch) or we need to run all initializers (else branch).

I'll take a look at moving those to a class / function.

Reply:
I think this would be equivalent:

```suggestion
                if (!sampledOut)
            {
                for (int index = 0; index < this.configuration.TelemetryInitializers.Count; index++)
                {
                    if (telemetryWithSampling?.SupportsProactiveSampling ?? false)
                    {
                        if (shoudlCheckSamplingScore && !string.IsNullOrEmpty(telemetry.Context.Operation.Id))
                        {
                            if (SamplingScoreGenerator.GetSamplingScore(telemetry.Context.Operation.Id) >= this.configuration.GetLastObservedSamplingPercentage(telemetryWithSampling.ItemTypeFlag))
                            {
                                telemetryWithSampling.IsProactivelySampledOut = true;
                                sampledOut = true;
                                break;
                            }
                            else
                            {
                                shoudlCheckSamplingScore = false;
                            }
                        }
                    }


                    try
                    {
                        this.configuration.TelemetryInitializers[index].Initialize(telemetry);
                    }
                    catch (Exception exception)
                    {
                        CoreEventSource.Log.LogError(string.Format(
                                                        CultureInfo.InvariantCulture,
                                                        "Exception while initializing {0}, exception message - {1}",
                                                        this.configuration.TelemetryInitializers[index].GetType().FullName,
                                                        exception));
                    }
                }
            }
```


This moves the `if (telemetryWithSampling?.SupportsProactiveSampling ?? false)` and the `if (shoudlCheckSamplingScore && !string.IsNullOrEmpty(telemetry.Context.Operation.Id))` both inside the `for` loop, which removes the need for the `Else`.

You could still reduce it further by moving the Sampling code into an external class/function.

Reply:
Ignore the last comment, I think this would be cleaner:

Can separate the sampling check from calling TelemetryInitializers.
Can remove an extra boolean.

```suggestion
            ISupportAdvancedSampling telemetryWithSampling = telemetry as ISupportAdvancedSampling;

            // Telemetry can be already sampled out if that decision was made before calling Track()
            bool sampledOut = telemetryWithSampling?.IsProactivelySampledOut ?? false;

            // If telemetry was not sampled out, we need to check its sampling score after we learn its Operation Id
            if (!sampledOut)
            {
                if (telemetryWithSampling?.SupportsProactiveSampling ?? false && !string.IsNullOrEmpty(telemetry.Context.Operation.Id))
                {
                    if (SamplingScoreGenerator.GetSamplingScore(telemetry.Context.Operation.Id) >= this.configuration.GetLastObservedSamplingPercentage(telemetryWithSampling.ItemTypeFlag))
                    {
                        telemetryWithSampling.IsProactivelySampledOut = true;
                        sampledOut = true;
                    }
                }
            }

            // Invoke TelemetryInitializers
            if (!sampledOut)
            {
                for (int index = 0; index < this.configuration.TelemetryInitializers.Count; index++)
                {
                    try
                    {
                        this.configuration.TelemetryInitializers[index].Initialize(telemetry);
                    }
                    catch (Exception exception)
                    {
                        CoreEventSource.Log.LogError(string.Format(
                                                        CultureInfo.InvariantCulture,
                                                        "Exception while initializing {0}, exception message - {1}",
                                                        this.configuration.TelemetryInitializers[index].GetType().FullName,
                                                        exception));
                    }
                }
            }
```

Reply:
Won't work unfortunately, we need `GetSamplingScore(telemetry.Context.Operation.Id)` after OperationCorrelationTelemetryInitializer was run and we cannot guarantee it's first in line and no one added other initiliazers before it. So this check is bound to be inside of the cycle. I used to have the implementation like you suggested about a couple of weeks ago and it didn't operate well.

If we separate Booleans while having this check inside the cycle, we will run this check on every step of the cycle if item is sampled it (check is expensive, so we can't). So, it means if one is not possible, the other is canceled out as well.